parsers/cpreprocessor.c

```diff
@@ -1244,54 +1300,380 @@ static void saveIgnoreToken(const char * ignoreToken, const char * param_name)
 	if(tokenEnd <= tokenBegin)
 		return;
 
+	cppMacroInfo * info = (cppMacroInfo *)eMalloc(sizeof(cppMacroInfo));
 
-	cppIgnoredTokenInfo * info = (cppIgnoredTokenInfo *)eMalloc(sizeof(cppIgnoredTokenInfo));
-
-	info->ignoreFollowingParenthesis = ignoreFollowingParenthesis;
-	info->replacement = replacement ? vStringNewInit(replacement) : NULL;
+	info->hasParameterList = ignoreFollowingParenthesis;
+	if(replacement)
+	{
+		cppMacroReplacementPartInfo * rep = (cppMacroReplacementPartInfo *)eMalloc(sizeof(cppMacroReplacementPartInfo));
+		rep->parameterIndex = -1;
+		rep->flags = 0;
+		rep->constant = vStringNewInit(replacement);
+		rep->next = NULL;
+		info->replacements = rep;
+	} else {
+		info->replacements = NULL;
+	}
 
-	hashTablePutItem(ignoreTokenTable,eStrndup(tokenBegin,tokenEnd - tokenBegin),info);
+	hashTablePutItem(defineMacroTable,eStrndup(tokenBegin,tokenEnd - tokenBegin),info);
 
 	verbose ("    ignore token: %s\n", ignoreToken);
 }
 
-static void freeIgnoredTokenInfo(cppIgnoredTokenInfo * info)
+static void saveMacro(const char * macro)
+{
+	CXX_DEBUG_ENTER_TEXT("Save macro %s",macro);
+
+	if(!macro)
+		return;
+
+	Assert (defineMacroTable);
+
+	const char * c = macro;
+	
+	// skip initial spaces
+	while(*c && isspacetab(*c))
+		c++;
+	
+	if(!*c)
+	{
+		CXX_DEBUG_LEAVE_TEXT("Bad empty macro definition");
+		return;
+	}
+	
+	if(!(isalpha(*c) || (*c == '_')))
+	{
+		CXX_DEBUG_LEAVE_TEXT("Macro does not start with an alphanumeric character");
+		return; // must be a sequence of letters and digits
+	}
+	
+	const char * identifierBegin = c;
+	
+	while(*c && (isalnum(*c) || (*c == '_')))
+		c++;
+	
+	const char * identifierEnd = c;
+
+#ifdef CXX_DO_DEBUGGING
+	char tmp = *c;
+	*(char *)c = 0;
+	CXX_DEBUG_PRINT("Macro identifier '%s'",identifierBegin);
+	*(char *)c = tmp;
```

Comment:
instead of messing with a constant string, I'd suggest
```C
CXX_DEBUG_PRINT("Macro identifier '%.*s'", identifierEnd - IdentifierBegin, identifierBegin`);
```